pkg/controllers/reconciler.go

```diff
@@ -208,30 +188,91 @@ func upgrade(ctx context.Context, r *reconciler, devicePlugin client.Object) {
 	}
 }
 
+// getAllDevicePluginObjects gets a list of untyped objects based on the parameters given to it.
+func getAllDevicePluginObjects(dynamic dynamic.Interface, ctx context.Context,
+	group string, version string, resource string, namespace string) (
+	[]map[string]interface{}, error) {
+	resID := schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resource,
+	}
+
+	list, err := dynamic.Resource(resID).Namespace(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	objList := make([]map[string]interface{}, len(list.Items))
+
+	for i, item := range list.Items {
+		objList[i] = item.Object
+	}
+
+	return objList, nil
+}
+
+// determinateSharedObjectReqs Determinates if the installed plugins require shared objects.
+// The result is one of three: no, may use and uses currently.
+func (r *reconciler) determinateSharedObjectReqs(ctx context.Context, req ctrl.Request) int {
+	ret := noSharedObjects
+
+	if !r.controller.PluginMayRequireSharedObjects() {
+		return ret
+	}
+
+	ret = mayUseSharedObjects
+
+	// Convert Kind into the resource name: GpuDevicePlugin -> gpudeviceplugins
+	resourcesName := strings.ToLower(r.pluginKind) + "s"
+
+	if r.dynamic == nil {
+		log.FromContext(ctx).Info("Cannot determinate shared object requests.")
+
+		return ret
+	}
+
+	// Get a list of untyped plugin objects.
+	objList, err := getAllDevicePluginObjects(r.dynamic, ctx, groupID, versionID, resourcesName, req.Namespace)
```

Comment:
is this just re-implementing the bookeper? I'm trying to understand where all the raw objects are needed just to define if the first daemonSet is also going to need the shared objects and if the last removed daemonSet is also going to need the shared objects removed.

Reply:
The old bookkeeper for me was counting how many CR objects were installed in the cluster. Its main user was the webhook that kept the cluster free from multiple CRs.
This new thing makes sure that serviceaccount and clusterrolebinging objects are created and deleted whenever there is need for them (or no need). It is quite a bit of extra complexity and its only user is GPU at the moment.
Other options would be:
1) Create serviceaccount&rolebinding per CR. Previously implemented, creates extra service account and rolebinding duplicates.
2) Create serviceaccount and rolebinding directly in the operator deployment. Operator doesn't create or delete them, but could check their existence possibly. The downside would be the unnecessary existence of the objects when no-one is using them (GPU plugin without resource management).

Reply:
I meant to ask if `r.controller.GetTotalObjectCount(ctx, r)` could be re-purposed for checking if the shared objects must be created

Reply:
ok, that had dropped from my context as I removed it in the initial patch. I'll see. Initially it would seem to be a possible alternative.

Reply:
I moved the CR list handling inside the controller thus removing the code from reconciler.

Reply:
thanks, looks better!